bip-0119.2.mediawiki

```diff
@@ -0,0 +1,184 @@
+<pre>
+  BIP: 119.2 (Pending Assignment)
+  Layer: Consensus (soft fork)
+  Title: CHECKTEMPLATEVERIFY V2
+  Author: Jeremy Rubin <j@rubin.io>
+  Comments-URI: https://github.com/bitcoin/bips/wiki/Comments:BIP-0119
+  Status: Draft
+  Type: Standards Track
+  Created: 2024-05-06
+  License: BSD-3-Clause
+</pre>
+
+==Abstract==
+
+This BIP proposes extensions to OP_CHECKTEMPLATEVERIFY BIP-119 as enhancements
+for LN Symmetry based constructions which use CTV and CSFS.
+
+
+==Summary==
+
+Adds a 20-byte CTV hash using HASH160 digest to save 12 bytes per script.  Adds
+a 1-byte flag (with no bits set) that allows for arbitrary data to be committed
+to inside the CTV hash.
+
+==Motivation==
+
+LN Symmetry prefers scripts as small as possible. Allowing the use of HASH160
+can allow users to reduce fees in exchange for minor loss of security.
+
+LN Symmetry also sometimes requires additional committed data for data
+availability purposes. When using BIP-118 APO, this extra data can be
+(although, according to BIP-341/342 should not be) placed in the Annex. As CTV
+does not commit to the annex, additional data can be committed to in an
+OP_RETURN.  However, OP_RETURNs are inefficient as they require 8 bytes of
+value plus a few VarInts to encode sizes, and use non discounted blockspace.
+Instead, committing to with CTV extra values on the stack can avoid these
+overheads.
+
+
+==Detailed Specification==
```

Comment:
One of the things I'm having trouble following here is precisely why this extra byte allows for Extra Data to be committed to. As far as I can tell the extra byte ensures that there are at least two stack elements but doesn't actually verify what those stack elements are.

I assume that CTV's natural semantics already result in committing to the contents of that second stack element but the mechanism is not obvious to me.

Reply:
this equation might help:

Sha256(CTVHASH(TX) || sha256(stack[-2])) == stack[-1][0:32]

more or less, check that the 32 bytes on the back of the stack match if you took the CTV hash and hashed it with some extra data.

The point is to enable signing stack[-1][0:32] with CSFS, which then also covers the extra data, allowing you to stuff arbitrary extra data into the witness.

Reply:
Ah yes now I see it. I got lost in the noise and didn't see the `-2` at the end [here](https://github.com/JeremyRubin/bips/blob/144d6179d6c2ff7145c9719c2e5db5f7a385a4dd/bip-0119.2.mediawiki?plain=1#L145).

I get it now. Thanks for explaining!